Bolt V2 with point type support

lutovich opened the pull request:
Introduced an extension of Bolt V1 with one additional type - point. New protocol version is exactly the same as the previous one but is able to serialize/deserialize points. 2D and 3D points are supported.

Made protocol test utilities able to work with different protocol versions.

lutovich committed:
Bolt V2 with point type support

Introduced an extension of Bolt V1 with one additional type - point.
New protocol version is exactly the same as the previous one but is
able to serialize/deserialize points. Arbitrary dimensional points
are supported. Serialization format is generic and allows `double[]`
array as point coordinate.

Made protocol test utilities able to work with different protocol
versions.

ali-ince reviewed (approved):
I think changes related to `Point` support looks good.

I had one point about `TransportTestUtil` or `Neo4jPack` usage in existing tests. I thought that it would be better if we inject both `Neo4jPackV1` and `Neo4jPackV2` as parameters into the test runs to ensure we do not break  V1 over V2 (at least for V2). WDYT?

lutovich committed:
Fix test compilation after rebase

lutovich committed:
Improve point handing in Bolt

 * introduced a new pack stream type - FLOAT_PAIR that represents two
   `double` values and contains just a single type header. It removes
   the need to encode two type headers for points with 2D coordinate

 * made 2D point serialization and deserialization code use new
   FLOAT_PAIR type

 * removed CRS table identifier from point serialization format
   because CRS can be uniquely identified by it's code

 * prohibited serialization/deserialization of points with more than
   2D coordinates

lutovich committed:
3D point support in bolt

Introduced a new pack stream type - FLOAT_TRIPLE that represents tree
`double` values and contains just a single type header. New 3D point
type uses FLOAT_TRIPLE to encode it's coordinate.

Fixed exception message to print correct hex value of the unexpected
type header. Previously it only contained first two hex digits, which
were always 'ff' for `byte` header values.

lutovich committed:
Parameterize bolt ITs with Neo4jPack versions

lutovich committed:
Moved bolt version into Neo4jPack & better test parameterization

`Neo4jPack` is currently the only versioned part of the bolt server
stack. This commit moves versioning into it and makes all ITs in the
bolt module use handshakes with parameterized versions.

lutovich committed:
Removed parametrization of BoltChannelAutoReadLimiterIT

It is a rather long running test that does not really care which
transport is used. This commit makes it only use plain socket
connection to reduce execution time.

lutovich commented:
@ali-ince made IT's in bolt module parameterized with `Neo4jPackV1` and `Neo4jPackV2`

lutovich committed:
Removed float pair and triple from Bolt

They were used for 2D and 3D points respectively. Instead points will
simply consist of 2 or 3 individually float values. This makes pack
stream code simpler. Also this optimization is not required in
absence of polygons.

lutovich committed:
Integration tests for 2D points in REST

lutovich committed:
Added tests for point arrays in Bolt and REST

ali-ince reviewed (approved):
:+1:
